src/NuGetGallery/Services/PackageUploadService.cs

```
@@ -18,19 +19,71 @@ public class PackageUploadService : IPackageUploadService
             _entitiesContext = entitiesContext ?? throw new ArgumentNullException(nameof(entitiesContext));
             _reservedNamespaceService = reservedNamespaceService ?? throw new ArgumentNullException(nameof(reservedNamespaceService));
             _validationService = validationService ?? throw new ArgumentNullException(nameof(validationService));
+            _config = config ?? throw new ArgumentNullException(nameof(config));
+        }
+
+        public async Task<PackageValidationResult> ValidatePackageWithKnownOwnerAsync(
+            PackageRegistration packageRegistration,
+            PackageArchiveReader nuGetPackage,
+            User owner)
+        {
+            var result = await ValidateSignatureFilePresenceAsync(packageRegistration, nuGetPackage, owner);
+            if (result != null)
+            {
+                return result;
+            }
+
+            return PackageValidationResult.Accepted();
+        }
+
+        private async Task<PackageValidationResult> ValidateSignatureFilePresenceAsync(
+            PackageRegistration packageRegistration,
+            PackageArchiveReader nugetPackage,
+            User owner)
+        {
+            if (packageRegistration == null)
+            {
+                // The package registration does not exist yet so we'll use a temporary one based off of the provided
+                // owner just for the purposes of signature file validation. Note that this entity is not added to the
+                // entity context.
+                packageRegistration = new PackageRegistration
+                {
+                    Owners = { owner },
+                };
+            }
+
+            if (await nugetPackage.IsSignedAsync(CancellationToken.None))
+            {
+                if (!packageRegistration.IsSigningAllowed()
+                    && _config.RejectSignedPackagesWithNoConfiguredCertificate)
```

check the config before method check for minor fast fail :)

---

Haha, love it. Good point. Fail faster-er.